File: ruby/bob/bob.rb
class Bob
  def self.hey(remark)
    @remark = remark

    return "Fine. Be that way!" if @remark.gsub(/\s/, '').length == 0
    return "Calm down, I know what I'm doing!" if uppercase? && question?  
    return "Sure." if question?
    
    uppercase? ? "Whoa, chill out!" : "Whatever."
  end

  private
  
  def self.uppercase?
    @remark.count("a-zA-Z") > 0 && @remark == @remark.upcase
  end
  
  def self.question?
    @remark.gsub(/\s/, '')[-1] == '?'
  end
end

module BookKeeping
  VERSION = 2 # Where the version number matches the one in the test.
end

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      empty?(input) -> "Fine. Be that way!"
      question?(input) && !(text?(input) && uppercase?(input)) -> "Sure."
      question?(input) && uppercase?(input) -> "Calm down, I know what I'm doing!"
      text?(input) && uppercase?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp text?(input) do
    String.upcase(input) != String.downcase(input)
  end

  defp empty?(input) do
    input |> String.trim() == ""
  end

  defp uppercase?(input) do
    String.upcase(input) == input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end
end
